lib/atmega328/render.cpp

```diff
@@ -0,0 +1,152 @@
+#include <stdint.h>
+#include "render.h"
+#include "util.h"
+#include "config.h"
+#include "fat_font.h"
+#include "max72xx_matrix.h"
+#include "millis.h"
+
+
+#define FONT_WIDTH 6
+#define FONT_HEIGHT 7
+
+void init_render()
+{
+    matrix_init();
+    init_millis();
+}
+
+static void draw_timers_indicator(const state_machine_t timers[])
+{
+    for (uint8_t i = 0; i < MAX_TIMERS; i++)
+    {
+        bool is_running = timers[i].state == RUNNING;
+        bool is_paused = timers[i].state == PAUSED;
+        bool is_ringing = timers[i].state == RINGING;
+
+        bool show_led = is_running || is_paused || is_ringing;
+
+        matrix_set_pixel(TIMERS_INDICATOR_COLUMN, i, show_led);
+    }
+
+}
+
+void blink_active_timer_indicator(uint8_t active_timer_index)
+{
+    static uint16_t last_blink_time;
+    uint16_t ms = millis();
+    static bool blink_state = true;
+    const bool should_toggle_active_timer_indicator = (ms - last_blink_time) >= ACTIVE_TIMER_INDICATOR_BLINK_RATE;
+    if (should_toggle_active_timer_indicator)
+    {
+        last_blink_time = ms;
+        blink_state = !blink_state;
+    }
+
+    for (uint8_t i = 0; i < MAX_TIMERS; i++)
+    {
+        bool is_active_timer = (i == active_timer_index);
+        if( is_active_timer)
+        {
+            matrix_set_pixel(TIMERS_INDICATOR_COLUMN, i, blink_state);
+        }        
+    }
+    matrix_update();
+}
+
+static void draw_digit(char digit, uint8_t x_offset, uint8_t y_offset, bool clear_digit)
+{
+    const uint8_t* ptr_digit = get_char(digit);
+    
+    for (uint8_t row = 0; row < FONT_HEIGHT; row++)
+    {
+        for (uint8_t col = 0; col < FONT_WIDTH; col++)
+        {
+            bool is_on = false;
+            if(!clear_digit)
+            {
+                is_on = ptr_digit[row] & bit(FONT_WIDTH - 1 - col);  // 6-bit wide
+            }
+            matrix_set_pixel(x_offset + col, y_offset + row, is_on);
+        }
+    }
+}
+
+static void draw_active_timer(uint16_t current_time, uint8_t x_offset, uint8_t y_offset, bool clear_active_timer)
+{
+    
+    uint8_t top_value, bottom_value;
+    if (current_time >= 3600) {
+        top_value = current_time / 3600;
+        bottom_value = (current_time / 60) % 60;
+    } else {
+        top_value = current_time / 60;
+        bottom_value = current_time % 60;
+    }
```

Comment:
Extracting a predicate can make the code easier to understand:

```suggestion
    const bool show_as_hh_mm = current_time >= 3600;
    if (show_as_hh_mm) {
        top_value = current_time / 3600;
        bottom_value = (current_time / 60) % 60;
    } else { // mm_ss
        top_value = current_time / 60;
        bottom_value = current_time % 60;
    }
```